### PR TITLE
stabilize const_collections_with_hasher and build_hasher_default_const_new

### LINKED ISSUE #102575: Tracking Issue for `const_collections_with_hasher`
Feature gate: `#![feature(const_collections_with_hasher)]`

This is a tracking issue for making [`HashMap::with_hasher`](https://doc.rust-lang.org/stable/std/collections/hash_map/struct.HashMap.html#method.with_hasher) and [`HashSet::with_hasher`](https://doc.rust-lang.org/stable/std/collections/hash_set/struct.HashSet.html#method.with_hasher) `const`.

### Public API

```rust
// std::collections
impl<K, V, S> HashMap<K, V, S> {
    pub const fn with_hasher(hash_builder: S) -> HashMap<K, V, S> { ... }
}

impl<T, S> HashSet<T, S> {
    pub const fn with_hasher(hasher: S) -> HashSet<T, S> { ... }
}
```

### Steps / History

- [x] Implementation: #102574
- [x] Final comment period (FCP)[^1]
- [ ] Stabilization PR (https://github.com/rust-lang/rust/pull/118427)

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### LINKED ISSUE #123197: Tracking Issue for fn const BuildHasherDefault::new()
Feature gate: `#![feature(build_hasher_default_const_new)]`

This is a tracking issue for a new `new` class method for BuildHasherDefault. It allows creating a BuildHasherDefault in const contexts.
Because HashMap::with_hasher constness is being stabilized in #118427 this will in turn allow creating empty `HashMap<K,V,BuildHasherDefault<H>>` in const context for any `H: Default + Hasher`.

### Public API

```rust
impl<H> BuildHasherDefault<H> {
    pub const fn new() -> Self;
}
```

### Steps / History

- [x] Implementation: #123198
- [x] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

@rustbot claim

### PR DESCRIPTION
**Pull request by RalfJung**
After a lot of preparatory work, finally we can stabilize creating `HashMap` in const context. :)

FCP for const_collections_with_hasher passed in https://github.com/rust-lang/rust/issues/102575.
Fixes https://github.com/rust-lang/rust/issues/102575.

FCP for build_hasher_default_const_new passed in https://github.com/rust-lang/rust/issues/123197.
Fixes https://github.com/rust-lang/rust/issues/123197.

Cc @Amanieu 
Release notes: https://github.com/rust-lang/rust/issues/133347

### CODE REVIEW

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval

File: src/tools/miri/cargo-miri/src/setup.rs [hunk 1]
```diff
@@ -136,7 +136,9 @@ pub fn setup(
     // But keep the overflow checks, they are cheap. This completely overwrites flags
     // the user might have set, which is consistent with normal `cargo build` that does
     // not apply `RUSTFLAGS` to the sysroot either.
-    let rustflags = &["-Cdebug-assertions=off", "-Coverflow-checks=on"];
+    // We also need to set `force-unstable-if-unmarked` to ensure the sysroot builds.
+    let rustflags =
+        &["-Cdebug-assertions=off", "-Coverflow-checks=on", "-Zforce-unstable-if-unmarked"];
```
Comment (RalfJung) [inline src/tools/miri/cargo-miri/src/setup.rs:141]: The recursive const stability logic relies on `-Zforce-unstable-if-unmarked`, so hashbrown needs to be built with that flag to ensure std can expose hashbrown `const fn` on stable.

Now I wonder if this won't be a problem for `-Zbuild-std`... does that set `-Zforce-unstable-if-unmarked`?

Comment (RalfJung) [reply src/tools/miri/cargo-miri/src/setup.rs:141]: Seems like the answer is yes, cargo does set that flag:

https://github.com/rust-lang/cargo/blob/668d82e3f0461ebbcb8eb2d86a8150be494a9427/src/cargo/core/compiler/mod.rs#L1214-L1223

Comment (RalfJung) [reply src/tools/miri/cargo-miri/src/setup.rs]: Turns out Miri also already sets that flag, the actual issue was setting in in boostrap.

File: library/std/src/collections/hash/map.rs [hunk 2]
```diff
@@ -204,6 +204,25 @@ use crate::ops::Index;
...
 /// }
 /// ```
+///
+/// # Usage in `const` and `static`
+///
+/// As explained above, `HashMap` is randomly seeded: each `HashMap` instance uses a different seed,
+/// which means that `HashMap::new` cannot be used in const context. To construct a `HashMap` in the
+/// initializer of a `const` or `static` item, you will have to use a different hasher that does not
+/// involve a random seed, as demonstrated in the following example. **`HashMap` constructed this
+/// way are not resistant against HashDoS!**
```
Comment (cuviper) [inline library/std/src/collections/hash/map.rs:214]: This should either be "`HashMap`s ... are not" or "A `HashMap` ... is not".

File: library/std/src/collections/hash/set.rs [hunk 3]
```diff
@@ -101,6 +101,25 @@ use crate::ops::{BitAnd, BitOr, BitXor, Sub};
...
 /// [`RefCell`]: crate::cell::RefCell
 /// [`Cell`]: crate::cell::Cell
+///
+/// # Usage in `const` and `static`
+///
+/// Like `HashMap`, `HashSet` is randomly seeded: each `HashSet` instance uses a different seed,
+/// which means that `HashSet::new` cannot be used in const context. To construct a `HashSet` in the
+/// initializer of a `const` or `static` item, you will have to use a different hasher that does not
+/// involve a random seed, as demonstrated in the following example. **`HashSet` constructed this
+/// way are not resistant against HashDoS!**
```
Comment (cuviper) [inline library/std/src/collections/hash/set.rs:111]: Same subject-verb disagreement.

**Review by cuviper — APPROVED** [review]
`r=me` after a couple grammar nits.

**Pull request merged**